<img src="mioti.png" style="height: 100px">
<center style="color:#888">Módulo Data Science in IoT<br/>Asignatura: Data preprocessing</center>
# Challenge S3: Preguntas sobre AirBnB

## Objetivos

El objetivo de este challenge es reponder algunas preguntas sobre un dataset real de AirBnb.

## Configuración del entorno

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Carga de los datos
Como en otras ocasiones cargaremos los datos de un fichero csv.

In [2]:
df = pd.read_csv('dataset_airbnb_madrid.csv')

Una vez cargados los datos debemos inspeccionarlos, antes de empezar nuestro análisis.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13335 entries, 0 to 13334
Data columns (total 95 columns):
id                                  13335 non-null int64
listing_url                         13335 non-null object
scrape_id                           13335 non-null int64
last_scraped                        13335 non-null object
name                                13335 non-null object
summary                             12846 non-null object
space                               9700 non-null object
description                         13329 non-null object
experiences_offered                 13335 non-null object
neighborhood_overview               8308 non-null object
notes                               5037 non-null object
transit                             8264 non-null object
access                              7546 non-null object
interaction                         7523 non-null object
house_rules                         8666 non-null object
thumbnail_url                 

In [4]:
df.head()

id                            listing_url       scrape_id  \
0   7830063   https://www.airbnb.com/rooms/7830063  20170407214119   
1   9898596   https://www.airbnb.com/rooms/9898596  20170407214119   
2  15334645  https://www.airbnb.com/rooms/15334645  20170407214119   
3   1307795   https://www.airbnb.com/rooms/1307795  20170407214119   
4  17410608  https://www.airbnb.com/rooms/17410608  20170407214119   

  last_scraped                                          name  \
0   2017-04-08                     Quiet room in Plaza Mayor   
1   2017-04-08       Homely apartment in the heart of Madrid   
2   2017-04-08   Piso Muy Luminoso en pleno centro de Madrid   
3   2017-04-08              Rent room in the heart of Madrid   
4   2017-04-08  Luxury duplex penthouse in historic building   

                                             summary  \
0  Room in magnificent property in the historic c...   
1  Spacious apartment for up to 10 people, with a...   
2  Lugares de interés: Casa Lucio, Cine Doré, Cal...   
3                                                NaN   
4  Amazing duplex penthouse in a historic buildin...   

                                               space  \
0  Based in the historic center of Madrid, to 1 m...   
1  This spacious apartment is located in the cent...   
2  Lo mejor es su ubicación en pleno centro de Ma...   
3  Very nice room , whith private bath , in a bea...   
4  It is located in an emblematic building with e...   

                                         description experiences_offered  \
0  Room in magnificent property in the historic c...                none   
1  Spacious apartment for up to 10 people, with a...                none   
2  Lugares de interés: Casa Lucio, Cine Doré, Cal...                none   
3  Very nice room , whith private bath , in a bea...                none   
4  Amazing duplex penthouse in a historic buildin...                none   

                               neighborhood_overview        ...          \
0  In the historic heart of Madrid, near the Plaz...        ...           
1  The Living Santa Ana Apartments are located in...        ...           
2  La zona de la Latina es un área cultural y un ...        ...           
3                                                NaN        ...           
4  "Madrid de los Austrias "is a beautiful neighb...        ...           

  review_scores_value requires_license license jurisdiction_names  \
0                 9.0                t     NaN                NaN   
1                 9.0                t     NaN                NaN   
2                 9.0                t     NaN                NaN   
3                10.0                t     NaN                NaN   
4                10.0                t     NaN                NaN   

  instant_bookable cancellation_policy require_guest_profile_picture  \
0                t            moderate                             f   
1                t            moderate                             f   
2                t              strict                             f   
3                f              strict                             f   
4                f              strict                             f   

  require_guest_phone_verification calculated_host_listings_count  \
0                                f                              7   
1                                f                              1   
2                                f                              1   
3                                f                              1   
4                                f                             13   

   reviews_per_month  
0               5.04  
1               4.39  
2               5.57  
3               0.24  
4               4.29  

[5 rows x 95 columns]

## ¿Cuántos alojamientos tienen fotos?

In [5]:
df['picture_url'].isnull().value_counts()
# 13335 valores; contando los valores False con la función isnul()

False    13335
Name: picture_url, dtype: int64

13335 alojamientos tienen fotos

## ¿Cuál es el porcentaje de originalidad de las descripciones?.

In [6]:
len(df['description'].drop_duplicates())/len(df)
#se eliminan las filas que sean iguales con respecto al campo 'description', con la funcion drop duplicates. 

0.9654293213348332

96.54% --> es el porcentaje de orginalidad; usando drop_dupliicates para mantener la originalidad en la columna.

## ¿Cuántos alojamientos tiene de media cada dueño?

In [7]:
x = df['host_id'].drop_duplicates()
x
len(df['host_id'])/len(x)



1.6288017588860388

1.63 alojamientos de media

## ¿Cuáles son los 10 dueños que más alojamientos tienen?

In [8]:
x = df[~df['host_id'].duplicated()]
x = x[['host_id', 'host_total_listings_count']]
x.sort_values(by='host_total_listings_count', ascending=False).head(10)

host_id  host_total_listings_count
11246  116508614                      519.0
1339   107246527                      265.0
77      54310958                      207.0
120     21514948                      164.0
74       3566146                      142.0
8148    85138877                      136.0
45      15258781                      114.0
6311     4200246                       81.0
514     32046323                       68.0
11243  115528688                       63.0

In [9]:
print(x.sort_values(by='host_total_listings_count', ascending=False).head(10))

         host_id  host_total_listings_count
11246  116508614                      519.0
1339   107246527                      265.0
77      54310958                      207.0
120     21514948                      164.0
74       3566146                      142.0
8148    85138877                      136.0
45      15258781                      114.0
6311     4200246                       81.0
514     32046323                       68.0
11243  115528688                       63.0


## ¿Cuántos alojamientos tiene Airbnb en Madrid?
Tip: te recomiento que borres todos los alojamientos que no estén en Madrid y luego los cuentes.

In [10]:
y = df[~(df['city']=='Madrid')]
#Se crea un dataframe con todos los elementos fuera de Madrid para calcular con la total en la siguiente resta.
len(df)-len(y)

13191

13191 alojamientos tiene Airbnb en Madrid

## ¿Cuántos alojamientos requieren licencia?

In [11]:
len(df[df['requires_license']=='t'])

13333

13333 requieren licencia

## ¿Cuántos alojamientos que requieren licencia realmente la tienen?

In [12]:
#Se resta a la cuenta de las que requieren licencia la cuenta de las que no tienen con la suma de isnull()
z = df[df['requires_license']=='t']
len(z) - z.license.isnull().sum()

249

249 alojamientos que requieren licencia realmente la tienen

## ¿Cuántos alojamientos que requieren licencia realmente no la tienen?

In [13]:
z.license.isnull().sum()

13084

Con la funcion isnull() y sumando los elementos sobre la columna license, se observa que 13084 alojamientos que
requieren licencia, realmente no la tienen. 

## ¿Cual es el código postal con precio medio más caro de madrid?
Tip: excluye zipcodes atípicos

In [14]:
zipcode = df

zipcode['price'] = zipcode['price'].str.replace("$", "")
zipcode['price'] = zipcode['price'].str.replace(',', '')
zipcode['price'] = zipcode['price'].str.replace(' ', '')
zipcode = df.astype({"price": float})
zipcode.fillna('zipcode')
zipcode = zipcode.astype({"zipcode": str})

#Después de asegurar la columna price y zipcode como string, se escoge todas las strings que contengan 28
zipcode = zipcode[zipcode.zipcode.str.contains("28")]

#Se realiza la media y se saca el zipcode del valor más alto
final = zipcode.groupby('zipcode').agg({'price':'mean'})

final.sort_values(by='price', ascending=False).head(1)

price
zipcode       
28850    250.0

28850 - Torrejón de Ardoz tiene el apartamento con el precio medio más caro de Madrid.

## ¿En que rango se mueven los precios de los alojamientos en madrid?

In [15]:
zipcode.price.describe()

count    12877.000000
mean        68.901297
std        100.901159
min          9.000000
25%         31.000000
50%         53.000000
75%         80.000000
max       7700.000000
Name: price, dtype: float64

En la desripción del dataset se observa el rango entre 9 y 7700 Euros. 

## (extra) ¿Cuales son los factores que más afectan al precio?

In [16]:
zipcode.sort_values(by='price').head(100)

id                            listing_url       scrape_id  \
10309   4652573   https://www.airbnb.com/rooms/4652573  20170407214119   
13298  14845645  https://www.airbnb.com/rooms/14845645  20170407214119   
12063  17741678  https://www.airbnb.com/rooms/17741678  20170407214119   
12944  14305864  https://www.airbnb.com/rooms/14305864  20170407214119   
7646   13892058  https://www.airbnb.com/rooms/13892058  20170407214119   
9679   17494953  https://www.airbnb.com/rooms/17494953  20170407214119   
12933   1028662   https://www.airbnb.com/rooms/1028662  20170407214119   
1415    7522424   https://www.airbnb.com/rooms/7522424  20170407214119   
2039   16182884  https://www.airbnb.com/rooms/16182884  20170407214119   
12920  15838628  https://www.airbnb.com/rooms/15838628  20170407214119   
11683  11887620  https://www.airbnb.com/rooms/11887620  20170407214119   
12253   7892449   https://www.airbnb.com/rooms/7892449  20170407214119   
8745   13777565  https://www.airbnb.com/rooms/13777565  20170407214119   
12262  12322113  https://www.airbnb.com/rooms/12322113  20170407214119   
11678  14840572  https://www.airbnb.com/rooms/14840572  20170407214119   
8632   12926111  https://www.airbnb.com/rooms/12926111  20170407214119   
11663  13310604  https://www.airbnb.com/rooms/13310604  20170407214119   
11692  14823674  https://www.airbnb.com/rooms/14823674  20170407214119   
12309   8796971   https://www.airbnb.com/rooms/8796971  20170407214119   
12013  14811421  https://www.airbnb.com/rooms/14811421  20170407214119   
11987  14822814  https://www.airbnb.com/rooms/14822814  20170407214119   
13029  17445702  https://www.airbnb.com/rooms/17445702  20170407214119   
13030  16099561  https://www.airbnb.com/rooms/16099561  20170407214119   
13031  15673502  https://www.airbnb.com/rooms/15673502  20170407214119   
13043  17210787  https://www.airbnb.com/rooms/17210787  20170407214119   
13053  17210902  https://www.airbnb.com/rooms/17210902  20170407214119   
10710  17123595  https://www.airbnb.com/rooms/17123595  20170407214119   
13015  17606703  https://www.airbnb.com/rooms/17606703  20170407214119   
12009  13772954  https://www.airbnb.com/rooms/13772954  20170407214119   
11941  17375296  https://www.airbnb.com/rooms/17375296  20170407214119   
...         ...                                    ...             ...   
10889  15200478  https://www.airbnb.com/rooms/15200478  20170407214119   
13025   8234621   https://www.airbnb.com/rooms/8234621  20170407214119   
10888  15200350  https://www.airbnb.com/rooms/15200350  20170407214119   
10788  15199777  https://www.airbnb.com/rooms/15199777  20170407214119   
10912  15200865  https://www.airbnb.com/rooms/15200865  20170407214119   
11938  17985486  https://www.airbnb.com/rooms/17985486  20170407214119   
11960  14245241  https://www.airbnb.com/rooms/14245241  20170407214119   
12110  15481550  https://www.airbnb.com/rooms/15481550  20170407214119   
12120  16986163  https://www.airbnb.com/rooms/16986163  20170407214119   
12097   9807907   https://www.airbnb.com/rooms/9807907  20170407214119   
7539   15172799  https://www.airbnb.com/rooms/15172799  20170407214119   
11534   5650391   https://www.airbnb.com/rooms/5650391  20170407214119   
13123  15695879  https://www.airbnb.com/rooms/15695879  20170407214119   
7936    1292772   https://www.airbnb.com/rooms/1292772  20170407214119   
9904   11840998  https://www.airbnb.com/rooms/11840998  20170407214119   
11538   5650684   https://www.airbnb.com/rooms/5650684  20170407214119   
13179  17993110  https://www.airbnb.com/rooms/17993110  20170407214119   
12035  15475748  https://www.airbnb.com/rooms/15475748  20170407214119   
2694   17638103  https://www.airbnb.com/rooms/17638103  20170407214119   
11475   6606510   https://www.airbnb.com/rooms/6606510  20170407214119   
12007   8776371   https://www.airbnb.com/rooms/8776371  20170407214119   
13207  16284278  https://www.airbnb.com/rooms/16284278  20170407214119   
12202  11365268  

Los factores principales que pueden afectar son la localización del apartamento, así como aspectos como el Scoring alto o la cancellation_policy flexible o moderate.

## Conclusiones

* ¿Qué opinas de AirBnB en Madrid?

In [17]:
zipcode['price'].sort_values(ascending=False).head(500)


10828    7700.0
8002     3000.0
7993     3000.0
8855     1890.0
2246     1200.0
10667    1000.0
5814     1000.0
957      1000.0
9190      875.0
12795     850.0
7626      800.0
11257     750.0
9127      750.0
10732     722.0
13080     722.0
11221     716.0
9339      716.0
9437      715.0
11349     712.0
11105     707.0
11214     706.0
10439     700.0
13045     689.0
9114      680.0
11332     650.0
3641      600.0
8154      600.0
8042      595.0
4368      595.0
10649     590.0
          ...  
10156     190.0
129       190.0
5105      190.0
3528      189.0
4517      189.0
13103     188.0
223       186.0
4153      186.0
10345     186.0
3286      186.0
4037      185.0
4005      185.0
5785      185.0
8245      185.0
10437     185.0
3825      185.0
3895      185.0
10541     185.0
2194      185.0
6735      183.0
9586      182.0
514       181.0
9738      180.0
10705     180.0
3081      180.0
2288      180.0
9191      180.0
10747     180.0
6607      180.0
13258     180.0
Name: price, Length: 500

In [18]:
airbnb = zipcode.sort_values(by='price', ascending=False ).head(100)
print(airbnb['zipcode'])

10828    28008
8002     28007
7993     28007
8855     28006
2246     28005
10667    28003
957      28013
5814     28013
9190     28016
12795    28042
7626     28007
9127     28016
11257    28035
13080    28051
10732    28035
9339     28039
11221    28040
9437     28020
11349    28023
11105    28034
11214    28049
10439    28001
13045    28041
9114     28016
11332    28023
3641     28014
8154     28001
8042     28001
4368     28004
10649    28003
         ...  
11220    28035
11250    28035
11207    28034
9231     28016
13289    28042
11335    28023
11330    28023
10731    28035
13187    28037
5763     28004
11218    28035
11239    28040
10704    28035
10925    28008
9672     28039
11343    28023
10688    28035
10690    28035
11265    28035
11235    28040
10693    28035
11188    28011
10734    28035
10717    28035
9375     28020
10737    28035
10686    28035
8935     28006
11226    28040
8971     28002
Name: zipcode, Length: 100, dtype: object


En general, se puede decir que los precios por noche se elevan para casos como habitaciones privadas y, especialmente, en el centro de Madrid y sus alrededores. Haciendo una búsqueda rápida se observa que los 500 pisos mas caro oscilan con un precio de 180 - 7000 Euros por noche, y el Codigo Postal 280xx indica que es alrededor del centro de la ciudad.